### ISSUE TITLE
i didn't understand preprocess function

### ISSUE THREAD

**Issue by cestcode** [phkhanhtrinh23/inverse_text_normalization#1]
Thanks for this repo! Can you explain [this preprocess function](https://github.com/phkhanhtrinh23/inverse_text_normalization/blob/dd52f004d235333e7763a47b2a11d94dffba6e8c/preprocess.py#L146)?

**Comment by phkhanhtrinh23** [comment]
This function, `preprocess_function`, prepares input data for a natural language processing (NLP) task, likely related to token classification or sequence-to-sequence tasks. Here's a detailed breakdown of the code:

---

### **Function Purpose**
The function processes a batch of input data to:
1. Tokenize source (`src`) and target (`tgt`) sequences.
2. Generate features for training or inference, including token IDs, attention masks, and additional labels for specific tasks like spoken normalization.

### **Step-by-Step Explanation**

---

#### **Initialization**
1. **Global Tokenizer:**
   ```python
   global tokenizer
   if tokenizer is None:
       tokenizer = model.init_tokenizer()
   ```
   - The function uses a global `tokenizer` object for tokenizing text.
   - If `tokenizer` is not initialized, it calls `model.init_tokenizer()`.

---

#### **Feature Extraction Loop**
2. **Loop through `src` and `tgt` pairs:**
   ```python
   for src_words, tgt_words in zip(batch["src"], batch["tgt"]):
   ```
   - The function processes each pair of source (`src_words`) and target (`tgt_words`) sentences in the batch.

3. **Initialize Variables for Each Pair:**
   ```python
   src_ids, pad_ids, src_lengths, tgt_ids, tgt_lengths = [], [], [], [], []
   src_spoken_label = []  
   src_spoken_idx = []
   tgt_spoken_ids = []
   ```
   - These lists store token IDs, attention masks, lengths of tokens, spoken normalization labels, indices of tokens requiring normalization, and token IDs of the normalized targets.

---

#### **Tokenization and Labeling**
4. **Check if Tokens are Identical:**
   ```python
   is_remain = False
   if src == tgt:
       is_remain = True
   ```
   - If `src` and `tgt` are identical, it sets `is_remain` to `True`, indicating no normalization is needed.

5. **Tokenize Source:**
   ```python
   src_tokenized = tokenizer(src)
   ```
   - The source sentence is tokenized into IDs using the `tokenizer`.

6. **Handle Special Cases (e.g., Email):**
   ```python
   if validators.email(tgt):
       tgt_tokenized = tokenizer(tgt.replace('@', ' @'))
   else:
       tgt_tokenized = tokenizer(tgt)
   ```
   - Emails are processed with a special hardcoded fix where `@` is separated to ensure proper tokenization.

7. **Assign Spoken Labels:**
   ```python
   if is_remain:
       src_spoken_label.extend([0 if random.random() < 0.5 else -100 ...])
   else:
       src_spoken_label.extend([1] + [2] * (len(src_tokenized["input_ids"][1:-1]) - 1))
   ```
   - If `src` equals `tgt`, the label is `0` with some chance of being ignored (`-100`).
   - Otherwise, the label is `1` for the first token and `2` for subsequent tokens.

---

#### **Attention Mask and Lengths**
8. **Extract Attention Mask and Lengths:**
   ```python
   pad_ids.extend(src_tokenized["attention_mask"][1:-1])
   src_lengths.append(len(src_tokenized["input_ids"]) - 2)
   tgt_lengths.append(len(tgt_tokenized["input_ids"]) - 2)
   ```
   - The attention mask (`pad_ids`) identifies which tokens are real and which are padding.
   - Lengths of the tokenized sequences are calculated for both `src` and `tgt`.

---

#### **Sample Filtering**
9. **Discard Samples That Are Too Long:**
   ```python
   if len(src_ids) > 80 or len(tgt_ids) > 80:
       break
   ```
   - Any sequence exceeding a length of 80 tokens is ignored.

10. **Ensure Valid Sequences:**
   ```python
   if len(src_ids) < 1 or len(tgt_ids) < 1:
       continue
   ```
   - Skip sequences with insufficient tokens.

---

#### **Collect Features**
11. **Append Processed Features:**
   ```python
   features.append({
       "input_ids": src_ids,
       "attention_mask": pad_ids,
       "spoken_label": src_spoken_label,
       "inputs_length": src_lengths,
       "outputs": tgt_ids,
       "outputs_length": tgt_lengths,
       "src_spoken_idx": src_spoken_idx,
       "tgt_spoken_ids": tgt_spoken_ids
   })
   ```
   - Each processed sample is added to the `features` list. This dictionary includes:
     - `input_ids`: Tokenized IDs of `src`.
     - `attention_mask`: Attention mask for `src`.
     - `spoken_label`: Labels indicating spoken normalization requirements.
     - `inputs_length` and `outputs_length`: Lengths of tokenized sequences.
     - `outputs`: Tokenized IDs of `tgt`.
     - `src_spoken_idx` and `tgt_spoken_ids`: Indices and target IDs for normalization.

---

### **Output**
12. **Return Processed Features:**
   ```python
   return features
   ```
   - The function returns a list of dictionaries, each containing preprocessed data for a `src`-`tgt` pair.

---

### **Summary**
This function prepares data for a task involving tokenization and spoken normalization. It:
- Tokenizes and aligns `src` and `tgt` sequences.
- Identifies tokens requiring spoken normalization.
- Generates additional features like attention masks, sequence lengths, and normalization labels.
- Ensures data quality by filtering invalid or overly long samples.
